lib/ask/ecto_types/schedule.ex

```diff
@@ -225,21 +262,39 @@ defmodule Ask.Schedule do
     Timex.Timezone.resolve(schedule.timezone, {erl_date, erl_time})
   end
 
-  defp next_available_date_time_internal(schedule, erl_date) do
-    erl_date = next_available_date(schedule, erl_date)
-    at_start_time(schedule, erl_date)
+  defp at_end_time_erl(schedule, erl_date) do
+    erl_time = schedule.end_time |> Time.to_erl
+    Timex.Timezone.resolve(schedule.timezone, {erl_date, erl_time})
   end
 
-  defp next_available_date(schedule, erl_date) do
-    erl_date = Timex.shift(erl_date, days: 1)
-    if day_of_week_available?(schedule, erl_date) do
+  defp next_available_date(schedule, erl_date, backward, limit) do
+    if limit_exceed(erl_date, backward, limit) do
       erl_date
     else
-      next_available_date(schedule, erl_date)
+      shift_days = if backward, do: -1, else: 1
+      erl_date = Timex.shift(erl_date, days: shift_days)
+      if day_of_week_available?(schedule, erl_date) do
+        erl_date
+      else
+        next_available_date(schedule, erl_date, backward, limit)
+      end
+    end
+  end
+
+  defp limit_exceed(_date_time, _backward, nil = _limit), do: false
```

Comment:
Not mandatory (because is just naming), but I would change it to `date_exceeds_limit?` or `limit_exceeded?`